### PR TITLE
Refactor of GeohashCellQuery

### PR DESCRIPTION
**Pull request by colings86**
Moving the query building functionality from the parser to the builders
new toQuery() method analogous to other recent query refactorings.

Relates to #10217

PR goes against the query-refactoring branch

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java [hunk 1]
```diff
@@ -42,6 +43,8 @@
 
     /** Default for boost to apply to resulting Lucene query. Defaults to 1.0*/
     public static final float DEFAULT_BOOST = 1.0f;
+    public static final ParseField NAME_FIELD = new ParseField("_name");
+    public static final ParseField BOOST_FIELD = new ParseField("boost");
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: I think we have to decide whether ParseFields should be in query builders or query parsers. Good to share these two though given that they are always the same. They should indeed be in some base class.

File: core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java [hunk 2]
```diff
@@ -149,19 +161,73 @@ public Builder neighbors(boolean neighbors) {
...
+        public boolean neighbors() {
+            return neighbors;
+        }
+
         public Builder field(String field) {
             this.field = field;
             return this;
         }
 
+        public String field() {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java]: rename to fieldName, as we do in other queries?

File: core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java [hunk 3]
```diff
@@ -149,19 +161,73 @@ public Builder neighbors(boolean neighbors) {
...
+            return field;
+        }
+
         @Override
-        protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-            builder.startObject(NAME);
+        public QueryValidationException validate() {
+            QueryValidationException errors = new QueryValidationException();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java]: use QueryValidationException#addValidationError ?

File: core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java [hunk 4]
```diff
@@ -149,19 +161,73 @@ public Builder neighbors(boolean neighbors) {
...
+            if (!(fieldType instanceof GeoPointFieldMapper.GeoPointFieldType)) {
+                throw new QueryParsingException(context.parseContext(), "failed to parse [{}] query. field [{}] is not a geo_point field",
+                        NAME, field);
+            }
+
+            GeoPointFieldMapper.GeoPointFieldType geoFieldType = ((GeoPointFieldMapper.GeoPointFieldType) fieldType);
+            if (!geoFieldType.isGeohashPrefixEnabled()) {
+                throw new QueryParsingException(context.parseContext(),
+                        "failed to parse [{}] query. [geohash_prefix] is not enabled for field [{}]", NAME, field);
+            }
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java]: we only need to move to `QueryShardException`s here

**Comment by javanna** [comment]
left a few minor comments, looks good

**Comment by colings86** [comment]
@javanna I pushed a commit which addresses your comments

File: core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java [hunk 5]
```diff
@@ -93,9 +97,9 @@ public static Query create(QueryShardContext context, GeoPointFieldMapper.GeoPoi
         // because a transformation from a geohash to a point an back to the
         // geohash will extend the accuracy of the hash to max precision
         // i.e. by filing up with z's.
-        private String field;
+        private String fieldName;
         private String geohash;
-        private int levels = -1;
+        private Integer levels = null;
         private boolean neighbors;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java]: don't we have sensible defaults for levels and neighbors?

**Comment by javanna** [comment]
left one tiny comment around default values in the builder, LGTM besides that

File: core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java (hunk 5 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java]: I'll add a default of `false` for neighbors but `null` is actually the sensible default for `levels` as it means 'do not truncate the given geohash'. The `levels` parameter is only there to truncate the precision of the given geohash if required

Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/GeohashCellQuery.java]: sounds good, thanks

**Pull request merged**